0.20.0 release (#177)
TeamNewPipe/website: _posts/2020-09-15-newpipe-0.20.0.md

---
layout: post
title: "NewPipe 0.20.0 released: A new major version, a new unified player, a new notification experience... A lot is new this time!"
short: "NewPipe 0.20.0 released"
date: 2020-09-15 05:00:00 +0100
categories: [pinned, announcement]
author: Opusforlife & Poolitzer
image: newpipe
excerpt_separator: <!-- more -->
---

Hey there! Guess who is back? We are. Admittedly, we're quite a bit late with this release. But there is a _very_ good reason for that. Are you ready for it? Sit down, grab some popcorn and a beverage of your choice. This is going to be long.
<!-- more -->

## A new unified player

> "What in the name of the coronavirus is the _unified player_?" you ask. Well, first of all, that's a very weird expression. Where did you even get that one from... - you know what? Save it. We don't want to know. You just keep on using that, we'll keep on talking about the new player.

Reviewer: Why `...` instead of `…`?

Author: uhm, to show more hesitation I guess?

Reviewer 2: How do you even type the ellipsis in a simple text editor? In something like Office, three periods get replaced by it.

Reviewer: With most GUI toolkits on Linux, it's Ctrl+Shift+U 2026

A new player it indeed is. Up until now, NewPipe had three different players bundled in its code: the main, pop-up and background players. All three had separate code and separate playing queues. When you switched from one player to another, the stream would start from the beginning instead of carrying on from where you left. In the background, the old player and its queue closed itself, while the new one took over. If that sounds complicated and cumbersome, that's because it is. At least if you ask us.

Well, no more! Now, when you switch from one player to another, you resume _exactly_ where you left off, and even your playing queue carries over! Since there is only one player (although with all of the three different playing modes), there is only one queue as well. Ah, serendipity.

This was made possible by rewriting the code so that all players run on one common service. When you switch from one player to another, this service remains running, preserving the stream data, and the new player just picks up from where the old one left off.

### Video details page

> A more unified experience is not the only feature this change brought you. There is one which is immediately noticeable once you open a video in the new version: the video details page. Earlier, when you opened the video details, you were shown a thumbnail, video description, comments, etc. Tapping the thumbnail would open the player in full-screen, hiding the rest of these elements. Now, when you're in portrait mode, the video plays right in the video details page - just like YouTube! There is even a new 'Autoplay' setting which plays the video automatically when you open video details - also just like YouTube! This setting will be turned on by default. Of course, you can turn it off if you prefer the behaviour of the previous versions.

Reviewer: It's only turned on by default for Wi-Fi

### Video orientation

Previously, the main video player would always play videos in full-screen regardless of your selected orientation. This was fine for landscape videos in landscape mode and portrait videos in portrait mode, but it resulted in a lot of wasted space if the video's orientation didn't match the player's. No more! There is now a full-screen button which replaces the orientation switch button, and it is _smart_. It will detect if the video is landscape or portrait, and will ensure the player is in the correct orientation when you enter full-screen! How's _that_ for convenience and usability? Not only that, but if auto-rotate is enabled, and the video is a landscape one, the button will disappear because you don't need it! (In portrait mode, you still need the button to go into full-screen from the video details.)

### Minimized player

> You can now swipe down on the video in the video details page to _minimize_ it - just like Youtu... wait no. It's better. You get a minimized player at the bottom, and you can play/pause or even close it from there. How cool is that? The video will keep playing (or stay paused, if you prefer that) while you browse NewPipe at your leisure! No more having to finish a video before browsing mindlessly! Yay distractions! (In landscape mode, you can swipe down on the title to minimize as well. The swipe area is confined to that region because the rest is used for the existing volume and brightness gestures, as you know.)

Reviewer 3: YouTu`…`

Reviewer 2: Oops.

* * *

On a serious note, this was a massive, massive change to not only the code, but also the app's workflow, so this PR was tested, reviewed and modified lots of times, over a long period of time, before it was allowed to see the light of day with this release. A huge thank you to you, [@avently](https://github.com/avently), for having the patience and dedication to keep adding features and bug fixes until everyone was satisfied! A different person probably would have gone insane. And a big thank you to the maintainers for doggedly reviewing this gigantic PR repeatedly. Those ~6600 additions and ~5600 removals must have felt intimidating!

Naturally, since this was such a massive change, several bugs also slipped through. But [@avently](https://github.com/avently), with the help of [@blackbox87](https://github.com/blackbox87), [@Stypox](https://github.com/Stypox) and [@TobiGr](https://github.com/TobiGr) fixed them rapidly as they were reported. 

> ## The Notification is dead, long live the Notification!

Reviewer: Why is notification capitalized?

Author: because its our specific notification.

Reviewer 2: Resolved, @wb9688?

If you care about such things, you have likely noticed that NewPipe's notification looks quite different from most media player apps these days. It has a progress bar, buttons in odd places, and a permanent grey colour. This is because NewPipe uses an old API for notifications called RemoteView. This allowed for a lot of customisation, and that was good at the time it was in vogue. However, it has been deprecated and superseded by a more modern API.

> Enter the **Age of Mediastyle notification**:

Reviewer: `MediaStyle`

### Android 9

First of all, we say "sorry" to our users on Android 9 and below. The progress bar is gone. As we've said above, it will only be shown on Android 10 and above. We're a small team and can't maintain code for the old and new types of notifications simultaneously, so the old one has to go. And not just that; the old notification isn't even supported Android 11 onwards. Bye-bye RemoteView notifications. We shall miss you! :'(

But you get a lot in exchange for that sacrifice!

### New features

You get five CUSTOMISABLE - you read it right! - **CUSTOMISABLE** action buttons to control playback. There is a new notification menu in the apps settings, and you can choose which button should perform which player action. Play/pause, forward/rewind, previous/next, shuffle, repeat, close, or even nothing!

Then, you can choose which three of these five buttons show in the compact notification (remember that little arrow you use to expand/contract the notification?).

> For users on **Android 8+**: Colours! Warm colours. Cool colours. All the colours! As we've noted above, Mediastyle notifications adapt to the colours of the cover art, which looks just plain awesome.

Reviewer: Again, it's `MediaStyle`

> For users on **Android 10+**: Seekbar! The old notification had a progress bar, but that was just a visual indicator. Now you can actually seek from the notification itself. Marvel at the massive boost in convenience, and have fun seeking!

Reviewer: Why does the word after the colon get capitalized here but not later in this post? It should be consistent

Author: No, it can be both. Basically, if a colon replaces a dot, you capitalize, if it replaces a comma, you dont.

https://www.grammarly.com/blog/capitalization-after-colons/

Reviewer: Interesting… you should follow whatever is correct for British English (or at least, you're using e.g. "colour" in the rest of the article)

Reviewer 2: Oh, that was me. India has British spellings instead of American. I have no preference regarding that. Either way would be fine for me.

Author: I guess resolved?

It's 2020, and our default resolutions also realized this. Instead of having 360p everywhere, main player videos are opened in 720p at 60FPS now, and pop-up ones in 480p. Thank you, [@B0pol](https://github.com/B0pol). NewPipe now also remembers what kind of media you downloaded the last time, and will auto-select it when you hit download on a new video. This was done by [@vmazoyer](https://github.com/vmazoyer).

Since 'Autoplay' is a new configurable setting in this release, [@opusforlife2](https://github.com/opusforlife2) changed the 'Autoplay' toggle in the video details to say "Auto-queue" instead. Now it's clearer what that toggle does without leaving room for any confusion.

## Fixes

If you noticed that next to offline search input now working, the actual text of the suggestions look better now, you are correct. [@TobiGr](https://github.com/TobiGr) changed the colours used to the default ones.

An empty error report in certain kinds of crashes is not empty any more, since [@Stypox](https://github.com/Stypox) took care of properly calling code. Yay, code!

> Thanks to [@wb9688](https://github.com/wb9688), you can open a video in the local playlist tab, whereas earlier, the app would crash. And thanks to [@BoFFire](https://github.com/BoFFire), if you open the app in the Taqbaylit language, its called `Taqbaylit` now and not `Tamaziɣt Taqbaylit`.

Reviewer: Those aren't at all related and should have separate paragraphs, same thing below.

Author: Previously we put two "unrelated" things in a paragraph as well. I think having two changes per paragraph is a great tradeoff between overall post length and readability. If we would put just one in a paragraph, the length of each section would be too long; if we put more then two in one, users might have the feeling at looking at a wall of text.

That is just a general rule of thumb for me and the posts I write, it changes on a case by case basis ofc, but thats how I used to do it bakc in the day. You younglings.

If more people agree with "one relatable change per paragraph" I will be happy to oblige.

Reviewer 2: I think it's better to have a single change per paragraph as well. But if two changes are related to one another, then it's better to have them in the same paragraph.

Author: Im just gonna wait on a third opinion here :(

Reviewer: I agree with @opusforlife2, but having a bunch of one sentence paragraphs is also ugly… not sure if it's possible with the current changes to group (semi-)related things together

Reviewer: Btw that PR didn't only fix local playlist, but potentially other places as well.

Author: added that, the discussion what we put in a paragraph and what not isn't really done yet so I'm gonna leave this as unresolved

The license pop-up doesn't disappear when you change orientation now. This was introduced by [@nmurali94](https://github.com/nmurali94). And [@budde92](https://github.com/budde25) ensured that you can now delete a file you just downloaded and then try to re-download it without the app crashing. We are not sure why you'd want to do this, but now you can do so without any worries.

[@mhmdanas](https://github.com/mhmdanas) changed a code call because its origin was redundant. Yay less redundancy. And [@wb9688](https://github.com/wb9688) fixed a crash when a video wouldn't have any next videos but NewPipe wanted to add one to the queue since auto-play were enabled.

## Development

> Development. Where all the magic happens which keeps this app running. It saw some improvements in this release (Warning: Nerd talk ahead): A checkstyle rule was added so it would error when a local variable which could be `final`wasn't declared as such (by [@XiangRongLin](https://github.com/XiangRongLin)), numbers and letters are allowed in the app package id (by [@Stypox](https://github.com/Stypox)), contributors need to to discuss their contributions now prior to open a PR (by [@gkeegan](https://github.com/gkeegan)) and some depreciations as well as a little code cleanup happened (by [@TacoTheDank](https://github.com/TacoTheDank).

Reviewer: It's "Checkstyle" and you're missing a space after `final`. It's "application ID", not "app package id", and it's "deprecations". Please end your file with a newline.

Talking about sharing, when you share a live stream with someone, NewPipe doesn't add a timestamp to the URL any more. This was done by [@nmurali94](https://github.com/nmurali94). Thanks.

The header layout has been improved: the alignment on the sides is consistent now and longer headings can fit themselves on two lines instead of just one.

If you ever tried to search something and then lost connection to the internet, or just tried to search while being offline, you would have noticed that NewPipe didn't let you type your search term. Instead it kept on rudely interrupting you with a network error. That is fixed now. You can type all you want, and it will only fail if you actually hit enter or press the search button. Both of these changes were brought to you by [@Stypox](https://github.com/Stypox). Thanks!

> 	In order to save some of that valuable mobile data, you were already able to turn off video thumbnails and profile pictures for comments. Now even the black placeholders which replaced them are gone and comments can comfortably take up the space left behind. Thanks, [@4D17Y4](https://github.com/4D17Y4)!

Reviewer 2: ```suggestion
In order to save some of that valuable mobile data, you were already able to turn off video thumbnails and profile pictures for comments. Now even the black placeholders which replaced them are gone and comments can comfortably take up the space left behind. Thanks, [@4D17Y4](https://github.com/4D17Y4)!
```

[@TobiGr](https://github.com/TobiGr) brought two more improvements into this release: the setting to remember pop-up size and position has a less redundant description, and when you switch to another app, the main player now minimizes to background audio by default to allow for a more seamless transition. Swoosh.

It's 2020, and our default resolutions also realized this. Instead of having 360p everywhere, main player videos are opened in 720p at 60FPS now, and pop-up ones in 480p. Thank you, [@B0pol](https://github.com/B0pol). NewPipe now also remembers what kind of media you downloaded the last time, and will auto-select it when you hit download on a new video. This was done by [@vmazoyer](https://github.com/vmazoyer).

Since 'Autoplay' is a new configurable setting in this release, [@opusforlife2](https://github.com/opusforlife2) changed the 'Autoplay' toggle in the video details to say "Auto-queue" instead. Now it's clearer what that toggle does without leaving room for any confusion.

### More Translation updates and improvements

> Apart from the language improvements mentioned above, some other interesting localization changes happened in this release. [@comradekingu](https://github.com/comradekingu), a frequent contributor to and language checker of NewPipe and other Open Source projects, has once again taken a close look at our strings. There was, as always, a lot of room for improvement. Our wonderful translators rushed to help and immediately implemented his suggestions.

Reviewer 2: ```suggestion
Apart from the language improvements mentioned above, some other interesting localization changes happened in this release. [@comradekingu](https://github.com/comradekingu), a frequent contributor to and language checker of NewPipe and other open source projects, has once again taken a close look at our strings. There was, as always, a lot of room for improvement. Our wonderful translators rushed to help and immediately implemented his suggestions.
```

> We are also happy to provide **new localizations** in this release: Sardinian, Bengali, Javanese and Berber

Reviewer 3: i forgot pt-pt, will fix later.

Author: Added

> If you are missing a translation or think a phrase needs adjustment, head over to [Weblate](https://hosted.weblate.org/engage/newpipe/) and help us improving NewPip's translation even more!

Reviewer 2: ```suggestion
If you are missing a translation or think a phrase needs adjustment, head over to [Weblate](https://hosted.weblate.org/engage/newpipe/) and help us improve NewPipe's translation even more!
```

[@mhmdanas](https://github.com/mhmdanas) changed a code call because its origin was redundant. Yay for less redundancy. When auto-queue is enabled, Newpipe wants to add a stream to the queue the moment you open a video. But if there are no video suggestions on that page, it causes a crash. This has been fixed by [@wb9688](https://github.com/wb9688).

## Development

Development. Where all the magic which keeps this app running happens. It saw some improvements in this release: (Warning: Nerd talk ahead)

- A Checkstyle rule was added so it would show an error when a local variable which could be `final` wasn't declared as such. ([@XiangRongLin](https://github.com/XiangRongLin))

- The extractor will ignore OTF streams now, since they don't work with progressive HTTP and the extractor can't work with DASH just yet ([@wb9688](https://github.com/wb9688))

> - PrettyTime is updated to 4.0.6, so a workaround can be removed from NewPipe's code and a perfomance improvement is part of it now ([@B0pol](https://github.com/B0pol)

Reviewer 3: ```suggestion
- PrettyTime is updated to 4.0.6, so a workaround can be removed from NewPipe's code and a perfomance improvement is part of it now ([@B0pol](https://github.com/B0pol))
```

Reviewer: This is for users, you should say what problems the update fixes, not that a workaround could be removed from our code.

Reviewer 4: Plural for days, used in relative date (e.g. 4 days ago), is fixed in russian with this prettytime update

- Numbers and uppercase letters are now allowed in the application ID. ([@Stypox](https://github.com/Stypox))

- Contributors need to to discuss their contributions now prior to opening a PR. ([@gkeegan](https://github.com/gkeegan))
	
- Some deprecations as well as a lot of code cleanup happened. ([@TacoTheDank](https://github.com/TacoTheDank))

> - The code now prevents the YouTube website from accessing the available Java packages from NewPipe's parser. This removes a possible attack angle. ([@wb9688](https://github.com/wb9688))

Reviewer 2: It's the correct term, so we should use it.

```suggestion
- The code now prevents the YouTube website from accessing the available Java packages from NewPipe's parser. This removes a possible attack surface. ([@wb9688](https://github.com/wb9688))
```

Author: still think angle sounds better :(

Reviewer 2: Use official terms. We can't suddenly say "Pull suggestion" instead of "Pull request", can we? O_O

---
layout: post
title: "NewPipe 0.20.0 released: A new major version, a new unified player, a new notification experience... A lot is new this time!"
short: "NewPipe 0.20.0 released"
date: 2020-09-15 05:00:00 +0100
categories: [pinned, announcement, release]
author: Opusforlife & Poolitzer
image: newpipe
excerpt_separator: <!-- more -->
---

Hey there! Guess who is back? We are. Admittedly, we're quite a bit late with this release. But there is a _very_ good reason for that. Are you ready for it? Sit down, grab some popcorn and a beverage of your choice. This is going to be long.
<!-- more -->

## A new unified player

"What in the name of the coronavirus is the _unified player_?" you ask. Well, first of all, that's a very weird expression. Where did you even get that one from... - you know what? Save it. We don't want to know. You just keep on using that, we'll keep on talking about the new player.

> A new player it indeed is. Up until now, NewPipe had three different players bundled in its code: the main, pop-up and background players. All three had separate code and separate playing queues. When you switched from one player to another, the stream would start from the beginning instead of carrying on from where you left. In the background, the old player and its queue closed itself, while the new one took over. If that sounds complicated and cumbersome, that's because it is. At least if you ask us.

Reviewer 2: ```suggestion
A new player it indeed is. Until now, NewPipe had three different players bundled in its code: the main, pop-up and background players. All three had separate code and separate playing queues. When you switched from one player to another, the stream would start from the beginning instead of carrying on from where you left. In the background, the old player and its queue closed itself, while the new one took over. If that sounds complicated and cumbersome, that's because it is. At least if you ask us.
```

> A new player it indeed is. Up until now, NewPipe had three different players bundled in its code: the main, pop-up and background players. All three had separate code and separate playing queues. When you switched from one player to another, the stream would start from the beginning instead of carrying on from where you left. In the background, the old player and its queue closed itself, while the new one took over. If that sounds complicated and cumbersome, that's because it is. At least if you ask us.

Reviewer 2: > When you switched from one player to another, the stream would start from the beginning instead of carrying on from where you left.

@Poolitzer Turns out this is only correct if you use the action buttons below the thumbnail. If you use the buttons in the player UI drop down, you correctly resume from where you left off (I think avently was the one who mentioned it in the RC thread). What to do here?

I'm thinking:

"All three had separate code and separate playing queues. When you switched from one player to another, in the background, the old player and its queue closed itself, while a new player was created, along with a new queue, after which it took over. If that sounds complicated and cumbersome, that's because it is. Moreover, if you accidentally switched using the player buttons below the thumbnail instead of the ones in the drop down menu in the player, the stream would start from the beginning instead of carrying on from where you left, and you would lose your (possibly carefully built) queue."

And for the next paras, something like:

"Well, no more! The buttons in the player drop down have been removed and the action buttons below the thumbnail work like they're supposed to, now, so you're at no risk whatsoever of losing your playing position or queue. Whew!

This was made possible by rewriting the code so that all players run on one common service. When you switch from one player to another, this service remains running, preserving the stream data, and the new player just picks up from where the old one left off. Everything remains the same, including the queue. Ah, serendipity."

> A new player it indeed is. Up until now, NewPipe had three different players bundled in its code: the main, pop-up and background players. All three had separate code and separate playing queues. When you switched from one player to another, the stream would start from the beginning instead of carrying on from where you left. In the background, the old player and its queue closed itself, while the new one took over. If that sounds complicated and cumbersome, that's because it is. At least if you ask us.

Author: All three had separate code and separate playing queues. When you switched from one player to another, in the background, the old player and its queue closed itself, while a new player was created, along with a new queue, which took over then.

Sounds good?

- Some deprecations as well as a lot of code cleanup happened. ([@TacoTheDank](https://github.com/TacoTheDank))

- The code now prevents the YouTube website from accessing the available Java packages from NewPipe's parser. This removes a possible attack surface. ([@wb9688](https://github.com/wb9688))

## Where to get this brand-new version

> NewPipe notifies you about new versions, you can download them when you press the notification. An alternative is the [GitHub release page](https://github.com/TeamNewPipe/NewPipe/releases). If you use the [F-Droid app](https://f-droid.org/), it notifies you as well about an update for [NewPipe](https://f-droid.org/packages/org.schabi.newpipe/) - please keep in mind that it can take F-Droid a while to update their repository. If you have problems installing you may need to uninstall NewPipe then reinstall (make sure to backup data).

Reviewer 2: ```suggestion
NewPipe notifies you about new versions. You can download them when you press the notification. An alternative is the [GitHub release page](https://github.com/TeamNewPipe/NewPipe/releases). If you use the [F-Droid app](https://f-droid.org/), it notifies you as well about an update for [NewPipe](https://f-droid.org/packages/org.schabi.newpipe/) - please keep in mind that it can take F-Droid a while to update their repository. If you have problems installing you may need to uninstall NewPipe then reinstall (make sure to backup data).
```

If you already had NewPipe installed through F-Droid's repository, to get this version of NewPipe you can do one of the following:
- Wait for them to update or
- Switch to the NewPipe repository by following the [directions in the announcement]({% post_url 2020-05-30-f-droid-repo %}) (if you had previously installed NewPipe from GitHub releases you will not have to uninstall NewPipe to switch)

> Now that you've updated, please let us know what your experience of the latest release is, especially bugs in need of fixing. As usual, you can reach out to us via IRC ([#newpipe on freenode](https://webchat.freenode.net/?channels=newpipe)), [open issues on GitHub](https://github.com/TeamNewPipe/NewPipe/issues/new) or ideally use our built-in crash reporter to send us machine-readable issue reports. You can even [send in fixes](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md#bug-fixing) yourself. If you have any other questions feel free to send them in the comments here and someone will reply to you.

Reviewer 2: ```suggestion
Now that you've updated, please let us know what your experience of the latest release is, especially bugs in need of fixing. As usual, you can reach out to us via IRC ([#newpipe on freenode](https://webchat.freenode.net/?channels=newpipe)), [open issues on GitHub](https://github.com/TeamNewPipe/NewPipe/issues/new) or, ideally, use our built-in crash reporter to send us machine-readable issue reports. You can even [send in fixes](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md#bug-fixing) yourself. If you have any other questions feel free to send them in the comments here and someone will reply to you.
```
